src/main/java/com/artipie/api/perms/RestApiPermission.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * The MIT License (MIT) Copyright (c) 2020-2021 artipie.com
+ * https://github.com/artipie/artipie/LICENSE.txt
+ */
+package com.artipie.api.perms;
+
+import com.artipie.security.perms.Action;
+import com.artipie.security.perms.AdapterBasicPermission;
+import java.security.Permission;
+import java.security.PermissionCollection;
+import java.util.Collection;
+import java.util.Enumeration;
+import java.util.Objects;
+import java.util.StringJoiner;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Base class for rest api permissions.
+ * @since 0.30
+ */
+public abstract class RestApiPermission extends Permission {
+
+    /**
+     * Required serial.
+     */
+    private static final long serialVersionUID = -2910976571451236971L;
+
+    /**
+     * Wildcard symbol.
+     */
+    private static final String WILDCARD = "*";
+
+    /**
+     * Canonical action representation. Is initialized once on request
+     * by {@link AdapterBasicPermission#getActions()} method.
+     */
+    private String actions;
+
+    /**
+     * Action mask.
+     */
+    private final transient int mask;
+
+    /**
+     * Actions type.
+     */
+    private final transient ApiActions type;
+
+    /**
+     * True, is any action is allowed.
+     */
+    private final transient boolean any;
+
+    /**
+     * Ctor.
+     * @param name Permission implementation name
+     * @param mask Permission action mask
+     * @param type Actions type
+     */
+    protected RestApiPermission(final String name, final int mask, final ApiActions type) {
+        super(name);
+        this.mask = mask;
+        this.type = type;
+        this.any = this.mask == type.all().mask();
+    }
+
+    @Override
+    public final boolean implies(final Permission permission) {
+        final boolean res;
+        if (permission != null && permission.getClass() == this.getClass()) {
+            final RestApiPermission that = (RestApiPermission) permission;
+            res = (this.mask & that.mask) == that.mask && this.getName().equals(that.getName());
```

Comment:
Due to the name argument for the same class type always the same, there is no sense checking the `name` equality.